### ISSUE TITLE
So How Do You Actually Calculate The Fibonacci Numbers?

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#170]
So How Do You Actually Calculate The Fibonacci Numbers?<br>
<br>
https://ift.tt/1fXHNOO<br>
<br>
Sahand Saba<br>
<br>
<div>
<h2>Introduction</h2>
<p>If you are a programmer you are probably a bit sick of the Fibonacci numbers. Code to calculate them is the go-to example for all sorts of situations. This is mostly because the Fibonacci numbers provide one of the simplest examples of a recurrence, making them a good example for any time recursion is part of the discussion. This then makes them a good example for introducing dynamic programming as well. But is that how you would actually calculate them? And the answer is no. Turns out, neither the recursive solution nor the dynamic programming approach are ideal. And no, I am not talking about the closed formula that uses floating point operations to calculate the numbers either. This is a quick post going over the correct way of calculating them. The tl;dr is how to calculate <span>\(F_n\)</span> in <span>\(O(\log n)\)</span> time, without using any floating point operations. Before presenting the fast algorithm though, I will go through the more well-known solutions first, for the sake of comparison.</p>
<p>The code is intended for Python 3, so <tt>range</tt> instead of <tt>xrange</tt> is used, <tt>zip</tt> is expected to return an iterator, etc. All of it should still run in Python 2, but likely will not be as efficient.</p>
<p>Before we start, here's a quick definition reminder:</p>
<div>\begin{equation*} F_n = F_{n-1} + F_{n-2} \end{equation*}</div>
<p>and <span>\(F_1 = F_2 = 1\)</span>.</p>
</div>
<div>
<h2>Using The Closed Form Formula</h2>
<p>Deriving a closed formula for <span>\(F_n\)</span> is a simple exercise in solving homogeneous linear recurrences. I skip the details, but if you have never seen the derivation, it is worth <a href="http://en.wikipedia.org/wiki/Fibonacci_number#Closed-form_expression">looking at</a>. The basic idea is <em>assuming</em> that <span>\(F_n = x^n\)</span> for some <span>\(x\)</span>, and then solving for <span>\(x\)</span>. This means</p>
<div>\begin{equation*} x^n = F_n = F_{n-1} + F_{n-2} = x^{n-1} + x^{n-2}, \end{equation*}</div>
<p>which implies</p>
<div>\begin{equation*} x^n - x^{n-1} - x^{n-2} = 0. \end{equation*}</div>
<p>Factoring out a <span>\(x^{n-2}\)</span> gives</p>
<div>\begin{equation*} x^2 - x - 1 = 0. \end{equation*}</div>
<p>Solving this using the quadratic formula gives</p>
<div>\begin{equation*} x = \frac{1 \pm \sqrt{5}}{2} \end{equation*}</div>
<p>which is where the golden ratio <span>\(\phi = \frac{1 + \sqrt{5}}{2}\)</span> comes from. After putting in the initial values and a bit more work, the following formula can be derived:</p>
<div>\begin{equation*} F_n = \lfloor \frac{\phi^n}{\sqrt{5}} + \frac{1}{2} \rfloor. \end{equation*}</div>
<p>This can immediately be used to calculate <span>\(F_n\)</span>:</p>
<div>
<pre>
<span>from</span> <span>__future__</span> <span>import</span> <span>division</span>
<span>import</span> <span>math</span>

<span>def</span> <span>fib</span><span>(</span><span>n</span><span>):</span>
    <span>SQRT5</span> <span>=</span> <span>math</span><span>.</span><span>sqrt</span><span>(</span><span>5</span><span>)</span>
    <span>PHI</span> <span>=</span> <span>(</span><span>SQRT5</span> <span>+</span> <span>1</span><span>)</span> <span>/</span> <span>2</span>
    <span>return</span> <span>int</span><span>(</span><span>PHI</span> <span>**</span> <span>n</span> <span>/</span> <span>SQRT5</span> <span>+</span> <span>0.5</span><span>)</span>
</pre>
</div>
<dl>
<dt>The Good:</dt>
<dd>It's fast and simple for small <span>\(n\)</span>.</dd>
<dt>The Bad:</dt>
<dd>Requires floating point operations. For larger <span>\(n\)</span>, more accuracy in floating point operations will be needed. Also, it is no good if we are calculating <span>\(F_n\)</span> modulo a number.</dd>
<dt>The Ugly:</dt>
<dd>Using irrational numbers to calculate <span>\(F_n\)</span> is mathematically beautiful, but from a computer science perspective just ugly.</dd>
</dl>
</div>
<div>
<h2>Recursion</h2>
<p>Recursion is perhaps the most obvious solution, and one which you have likely already seen a billion times, most likely as the go-to example of recursion. But here it is again, for the sake of comparison later. We can do it in one line in Python:</p>
<div>
<pre>
<span>fib</span> <span>=</span> <span>lambda</span> <span>n</span><span>:</span> <span>fib</span><span>(</span><span>n</span> <span>-</span> <span>1</span><span>)</span> <span>+</span> <span>fib</span><span>(</span><span>n</span> <span>-</span> <span>2</span><span>)</span> <span>if</span> <span>n</span> <span>&gt;</span> <span>2</span> <span>else</span> <span>1</span>
</pre>
</div>
<dl>
<dt>The Good:</dt>
<dd>Extremely simple implementation, following the mathematical definition.</dd>
<dt>The Bad:</dt>
<dd>Exponential running time. Very slow for larger <span>\(n\)</span>.</dd>
<dt>The Ugly:</dt>
<dd>The stack trace after you get a recursion depth limit error.</dd>
</dl>
</div>
<div>
<h2>Memoization</h2>
<p>The recursive solution has a major issue: overlapping recursion subtrees. This is because when <tt>fib(n)</tt> is called, <tt><span>fib(n-1)</span></tt> and <tt><span>fib(n-2)</span></tt> are calculated. But <tt><span>fib(n-1)</span></tt> in turn will calculate <tt><span>fib(n-2)</span></tt> independently, so <tt><span>fib(n-2)</span></tt> is calculated twice. Continuing in this manner, it is easy to see that <tt><span>fib(n-3)</span></tt> is calculated three times, and so on. So a lot of overlap in the computation.</p>
<p>And here's where <em>memoization</em> comes in. Basic idea is simple: store the results so we don't have to calculate them again. Both time and space complexity of this solution are linear. In the solution below I use a dictionary to do the memoization but a simple list or array could also be used, and would likely be faster if the upper limit of <span>\(n\)</span> is known in advance.</p>
<div>
<pre>
<span>M</span> <span>=</span> <span>{</span><span>0</span><span>:</span> <span>0</span><span>,</span> <span>1</span><span>:</span> <span>1</span><span>}</span>

<span>def</span> <span>fib</span><span>(</span><span>n</span><span>):</span>
    <span>if</span> <span>n</span> <span>in</span> <span>M</span><span>:</span>
        <span>return</span> <span>M</span><span>[</span><span>n</span><span>]</span>
    <span>M</span><span>[</span><span>n</span><span>]</span> <span>=</span> <span>fib</span><span>(</span><span>n</span> <span>-</span> <span>1</span><span>)</span> <span>+</span> <span>fib</span><span>(</span><span>n</span> <span>-</span> <span>2</span><span>)</span>
    <span>return</span> <span>M</span><span>[</span><span>n</span><span>]</span>
</pre>
</div>
<p>(Python note: you can also do the above using a decorator, for example <tt>functools.lru_cache</tt>.)</p>
<dl>
<dt>The Good:</dt>
<dd>Easy to turn a recursive solution to a memoized solution. Will turn previous exponential time complexity to linear complexity, at the expense of more space.</dd>
<dt>The Bad:</dt>
<dd>High space complexity to store all previous results and overhead of recursive calls and lookups.</dd>
<dt>The Ugly:</dt>
<dd>Same as the recursive solution: the stack trace after you get a recursion depth limit error.</dd>
</dl>
</div>
<div>
<h2>Dynamic Programming</h2>
<p>With memoization, it becomes clear that we don't actually need all the previous values stored, just the last two. Also, instead of starting from <tt>fib(n)</tt> and working backwards recursively, we can start from <tt>fib(0)</tt> and move forward to reach <tt>fib(n)</tt>. This leads the code below. The running time is linear, but space complexity is now constant. In practise, performance will also be much better since recursive function calls and the branching necessary with the memoization solution are no longer needed. Plus, the code looks much simpler.</p>
<p>This solution is often used as an example of <em>dynamic programming</em>, and is quite well-known.</p>
<div>
<pre>
<span>def</span> <span>fib</span><span>(</span><span>n</span><span>):</span>
    <span>a</span> <span>=</span> <span>0</span>
    <span>b</span> <span>=</span> <span>1</span>
    <span>for</span> <span>__</span> <span>in</span> <span>range</span><span>(</span><span>n</span><span>):</span>
        <span>a</span><span>,</span> <span>b</span> <span>=</span> <span>b</span><span>,</span> <span>a</span> <span>+</span> <span>b</span>
    <span>return</span> <span>a</span>
</pre>
</div>
<dl>
<dt>The Good:</dt>
<dd>Fast for small <span>\(n\)</span>, and very simple code.</dd>
<dt>The Bad:</dt>
<dd>Still linear number of operations in <span>\(n\)</span>.</dd>
<dt>The Ugly:</dt>
<dd>Nothing really.</dd>
</dl>
</div>
<div>
<h2>Using Matrix Algebra</h2>
<p>And finally, the solution that seems to get the least press, but is the correct one in the sense that it has the best time and space complexity. It is also useful since it can be generalized to any homogeneous linear recurrence. The basic idea here is to use matrices. It is rather simple to see that</p>
<div>\begin{equation*} {F_n \choose F_{n-1}} = \begin{pmatrix} 1 &amp; 1 \\ 1 &amp; 0 \end{pmatrix} {F_{n-1} \choose F_{n-2}}. \end{equation*}</div>
<p>Repeated application of the above means that</p>
<div>\begin{equation*} \begin{pmatrix} 1 &amp; 1 \\ 1 &amp; 0 \end{pmatrix}^n = \begin{pmatrix} F_{n+1} &amp; F_n \\ F_n &amp; F_{n-1} \end{pmatrix}. \end{equation*}</div>
<p>(Side note: In fact, the two values for <span>\(x\)</span> that we got above, one of which was the golden ratio <span>\(\phi\)</span>, are the eigenvalues of the matrix above. So using the matrix equation above and a bit of linear algebra would be yet another way of deriving the closed formula.)</p>
<p>All right. So, why is this formulation useful? Because we can do exponentiation in logarithmic time. This is done using a method known as "exponentiation by squaring". I discuss it in a bit more detail in <a href="http://sahandsaba.com/cryptography-rsa-part-1.html">my post on RSA</a>, but the basic recursion is this:</p>
<div>\begin{equation*} A^m= \begin{cases} (A^{\frac{m}{2}})^2 &amp; \mbox{ if } m \mbox{ is even} \\ A \cdot (A^{\frac{m-1}{2}})^2 &amp; \mbox{ if } m \mbox{ is odd}. \end{cases} \end{equation*}</div>
<p>Other than that, we just need to be able to do matrix multiplication and we are set. This leads to the following code. Here I provide a recursive implementation of <tt>pow</tt>, since it is easier to understand. Feel free to look at the RSA post for an iterative version.</p>
<div>
<pre>
<span>def</span> <span>pow</span><span>(</span><span>x</span><span>,</span> <span>n</span><span>,</span> <span>I</span><span>,</span> <span>mult</span><span>):</span>
    <span>"""</span>
<span> Returns x to the power of n. Assumes I to be identity relative to the</span>
<span> multiplication given by mult, and n to be a positive integer.</span>
<span> """</span>
    <span>if</span> <span>n</span> <span>==</span> <span>0</span><span>:</span>
        <span>return</span> <span>I</span>
    <span>elif</span> <span>n</span> <span>==</span> <span>1</span><span>:</span>
        <span>return</span> <span>x</span>
    <span>else</span><span>:</span>
        <span>y</span> <span>=</span> <span>pow</span><span>(</span><span>x</span><span>,</span> <span>n</span> <span>//</span> <span>2</span><span>,</span> <span>I</span><span>,</span> <span>mult</span><span>)</span>
        <span>y</span> <span>=</span> <span>mult</span><span>(</span><span>y</span><span>,</span> <span>y</span><span>)</span>
        <span>if</span> <span>n</span> <span>%</span> <span>2</span><span>:</span>
            <span>y</span> <span>=</span> <span>mult</span><span>(</span><span>x</span><span>,</span> <span>y</span><span>)</span>
        <span>return</span> <span>y</span>

<span>def</span> <span>identity_matrix</span><span>(</span><span>n</span><span>):</span>
    <span>"""Returns the n by n identity matrix."""</span>
    <span>r</span> <span>=</span> <span>list</span><span>(</span><span>range</span><span>(</span><span>n</span><span>))</span>
    <span>return</span> <span>[[</span><span>1</span> <span>if</span> <span>i</span> <span>==</span> <span>j</span> <span>else</span> <span>0</span> <span>for</span> <span>i</span> <span>in</span> <span>r</span><span>]</span> <span>for</span> <span>j</span> <span>in</span> <span>r</span><span>]</span>

<span>def</span> <span>matrix_multiply</span><span>(</span><span>A</span><span>,</span> <span>B</span><span>):</span>
    <span>BT</span> <span>=</span> <span>list</span><span>(</span><span>zip</span><span>(</span><span>*</span><span>B</span><span>))</span>
    <span>return</span> <span>[[</span><span>sum</span><span>(</span><span>a</span> <span>*</span> <span>b</span>
                 <span>for</span> <span>a</span><span>,</span> <span>b</span> <span>in</span> <span>zip</span><span>(</span><span>row_a</span><span>,</span> <span>col_b</span><span>))</span>
            <span>for</span> <span>col_b</span> <span>in</span> <span>BT</span><span>]</span>
            <span>for</span> <span>row_a</span> <span>in</span> <span>A</span><span>]</span>

<span>def</span> <span>fib</span><span>(</span><span>n</span><span>):</span>
    <span>F</span> <span>=</span> <span>pow</span><span>([[</span><span>1</span><span>,</span> <span>1</span><span>],</span> <span>[</span><span>1</span><span>,</span> <span>0</span><span>]],</span> <span>n</span><span>,</span> <span>identity_matrix</span><span>(</span><span>2</span><span>),</span> <span>matrix_multiply</span><span>)</span>
    <span>return</span> <span>F</span><span>[</span><span>0</span><span>][</span><span>1</span><span>]</span>
</pre>
</div>
<dl>
<dt>The Good:</dt>
<dd>Constant space, logarithmic time.</dd>
<dt>The Bad:</dt>
<dd>Bit more complicated code.</dd>
<dt>The Ugly:</dt>
<dd>Having to deal with matrices, though I personally don't find them that bad.</dd>
</dl>
</div>
<div>
<h2>Performance Comparison</h2>
<p>The only two implementations that are worthy of testing are the dynamical programming one and the matrix based one. Thinking in terms of the number of digits of <span>\(n\)</span>, the matrix based solution is linear whereas the dynamical programming one is exponential, and that is more or less all that needs to be said. But to see it in practise, here is an example, calculating <tt>fib(10 ** 6)</tt> which has over two-hundred thousand digits.</p>
<div>
<pre>
n = 10 ** 6
Trying fib_matrix:      fib(n) has 208988 digits, took 0.24993 seconds.
Trying fib_dynamic:     fib(n) has 208988 digits, took 11.83377 seconds.
---
</pre>
</div>
</div>
<div>
<h2>A Graph Theoretic View</h2>
<p>This section is a bit of a tangent, and is not directly related to any of the code above. But I find it interesting, so here it is. Consider the graph below.</p>
<div><img alt="Directed graph generating the Fibonacci numbers." src="http://sahandsaba.com/images/fib_graph.png"></div>
<p>Now, count the number of paths of length <span>\(n\)</span> from <span>\(A\)</span> to <span>\(B\)</span>. For example, for length <span>\(n=1\)</span> we have just one path with label sequence <span>\(1\)</span>. For <span>\(n=2\)</span> we still have just one, but the label is now <span>\(01\)</span>. For <span>\(n=3\)</span> we have two with labels <span>\(001\)</span> and <span>\(101\)</span>.</p>
<p>It is a rather simple exercise to show that the number of paths of length <span>\(n\)</span> from <span>\(A\)</span> to <span>\(B\)</span> is in fact precisely <span>\(F_n\)</span>. If we write the adjacency matrix of the graph we get the matrix we used in the previous section. It is a well-known basic result in graph theory that given adjacency matrix <span>\(A\)</span>, the entries in <span>\(A^n\)</span> are the number of paths of length <span>\(n\)</span> in the graph. (This was <a href="http://www.math.unideb.hu/~ghorvath/publications/gwh2.pdf">one of the problems</a> in Good Will Hunting, by the way.) So we can provide yet another view of the above methods and formulations.</p>
<p>Why the labels on the edges? Well, turns out that if you consider the infinite sequence of symbols on doubly-infinite paths of the graph, you get what is known as a <a href="http://en.wikipedia.org/wiki/Subshift_of_finite_type">subshift of finite type</a> which is a type of a symbolic dynamical system. This particular subshift of finite type is known as the "golden mean shift", and is given by the set of forbidden words <span>\(\{11\}\)</span>. In other words, we get all doubly-infinite binary sequences with no two ones next to each other. The topological entropy of this dynamical system is the golden ratio <span>\(\phi\)</span>. It's interesting to see the number pop up in all sorts of places in mathematics.</p>
</div>
<br>
<br>
<br>
<br>
Python<br>
<br>
via Math &cup; Code http://sahandsaba.com/<br>
<br>
May 11, 2014 at 05:09AM
